dashboards/cls_dashboard.py

```diff
@@ -0,0 +1,159 @@
+    tokenizer: QuantizedMidiEncoder,
+    queries: list[str],
+    batch_size: int,
+    num_workers: int,
+):
+    dataset = load_dataset(dataset_name, split="validation")
+
+    ds = MidiDataset(dataset, quantizer, tokenizer, masking_probability=0.0)
+
+    idx = []
+
+    for query in queries:
+        idx_query = [i for i, name in enumerate(ds.dataset["source"]) if str.lower(query) in str.lower(name)]
+        idx += idx_query
+
+    ds = Subset(ds, indices=idx)
+
+    # dataloaders
+    dataloader = DataLoader(ds, batch_size=batch_size, num_workers=num_workers, shuffle=False)
+
+    return dataloader
+
+
+def plot_embeddings(X: np.ndarray, labels: list[str]):
+    pca = PCA(n_components=2)
+    X_reduced = pca.fit_transform(X)
+
+    fig = plt.figure(figsize=(10, 10))
+    sns.scatterplot(x=X_reduced[:, 0], y=X_reduced[:, 1], hue=labels)
+    st.pyplot(fig)
+
+def plot_confusion_matrix(cm: np.ndarray):
+    fig = plt.figure(figsize=(10, 10))
+    sns.heatmap(cm, annot=True)
+    st.pyplot(fig)
+
+
+def evaluate_classification(
+    model: RobertaForMaskedLM,
+    dataset_name: str,
+    quantizer: MidiQuantizer,
+    tokenizer: QuantizedMidiEncoder,
+    queries: list[str],
+    device: torch.device,
+):
+    dataloader = preprocess_dataset(
+        dataset_name,
+        quantizer=quantizer,
+        tokenizer=tokenizer,
+        queries=queries,
+        batch_size=1024,
+        num_workers=8,
+    )
+    query_to_id = {k: i for i, k in enumerate(queries)}
+
+    outputs = []
+    labels = []
+
+    # velocity time encoding
+    for batch in dataloader:
+        with torch.no_grad():
+            source = batch["source"]
+            input_token_ids = batch["input_token_ids"].to(device)
+
+            # shape [batch_size, seq_len, vocab_size]
+            out = model(input_ids=input_token_ids).logits
+            cls_embedding = out[:, 0, :]
+            outputs.append(cls_embedding)
+
+            for s in source:
+                for query in queries:
+                    if str.lower(query) in str.lower(s):
+                        labels.append(query)
+
+    # shape: [num_sequences, vocab_size]
+    X = torch.cat(outputs, dim=0).cpu().numpy()
+    y = np.array([query_to_id[q] for q in labels])
+
+    assert len(X) == len(y)
+
+    plot_embeddings(X, labels)
+
+    # reducing dimensions
+    pca = PCA(n_components=128)
+    X_reduced = pca.fit_transform(X)
+
+    X_train, X_test, y_train, y_test = train_test_split(X_reduced, y, test_size=0.2, random_state=42)
+
+    tree = DecisionTreeClassifier()
+    tree.fit(X_train, y_train)
+
+    acc = tree.score(X_test, y_test)
+    cm = confusion_matrix(tree.predict(X_test), y_test)
+    plot_confusion_matrix(cm)
+
+    return acc
+
+
+def main():
+    checkpoint = torch.load(
+        "checkpoints/masked-midi-modelling-2023-10-15-10-19-params-90.31M.ckpt"
+        # "checkpoints/pianomask_2023_10_08_19_10.pt"
+    )
+
+    cfg = checkpoint["config"]
+    device = cfg.train.device
```

Comment:
I think it would be better to map everything to CPU by default here, so it would be possible to run the dashboard while the model is training on the same machine 🤔